service/service_flash.go

```diff
@@ -0,0 +1,139 @@
+// modify or otherwise use the software for commercial activities involving the
+// Arduino software without disclosing the source code of your own applications.
+// To purchase a commercial license, send an email to license@arduino.cc.
+
+package service
+
+import (
+	"fmt"
+
+	"github.com/arduino/go-paths-helper"
+	"github.com/codeclysm/extract/v4"
+	"github.com/shirou/gopsutil/v4/disk"
+	"go.bug.st/downloader/v2"
+
+	"github.com/arduino/arduino-flasher-cli/internal/updater"
+	flasher "github.com/arduino/arduino-flasher-cli/rpc/cc/arduino/flasher/v1"
+)
+
+func (s *flasherServerImpl) Flash(req *flasher.FlashRequest, stream flasher.Flasher_FlashServer) error {
+	// Setup callback functions
+	var responseCallback func(*flasher.FlashResponse) error
+	if stream != nil {
+		syncSend := NewSynchronizedSend(stream.Send)
+		responseCallback = syncSend.Send
+	} else {
+		responseCallback = func(*flasher.FlashResponse) error { return nil }
+	}
+	ctx := stream.Context()
+	downloadCB := func(msg *flasher.DownloadProgress) {
+		_ = responseCallback(&flasher.FlashResponse{
+			Message: &flasher.FlashResponse_DownloadProgress{
+				DownloadProgress: &flasher.DownloadProgress{
+					Message: msg.GetMessage(),
+				},
+			},
+		})
+	}
+	extractCB := func(msg *flasher.TaskProgress) {
+		_ = responseCallback(&flasher.FlashResponse{
+			Message: &flasher.FlashResponse_ExtractionProgress{
+				ExtractionProgress: &flasher.TaskProgress{
+					Name:      msg.GetName(),
+					Message:   msg.GetMessage(),
+					Completed: msg.GetCompleted(),
+				},
+			},
+		})
+	}
+	flashCB := func(msg *flasher.TaskProgress) {
+		_ = responseCallback(&flasher.FlashResponse{
+			Message: &flasher.FlashResponse_FlashProgress{
+				FlashProgress: &flasher.TaskProgress{
+					Name:      msg.GetName(),
+					Message:   msg.GetMessage(),
+					Completed: msg.GetCompleted(),
+				},
+			},
+		})
+	}
+
+	// Check if there is enough free disk space before downloading and extracting an image
+	d, err := disk.Usage(req.TempPath)
+	if err != nil {
+		return err
+	}
+	if d.Free/updater.GiB < updater.DownloadDiskSpace {
+		return fmt.Errorf("download and extraction requires up to %d GiB of free space", updater.DownloadDiskSpace)
+	}
+
+	client := updater.NewClient()
+	manifest, err := client.GetInfoManifest(ctx)
+	if err != nil {
+		return err
+	}
+
+	var rel *updater.Release
+	if req.Version == "latest" || req.Version == manifest.Latest.Version {
+		rel = &manifest.Latest
+	} else {
+		for _, r := range manifest.Releases {
+			if req.Version == r.Version {
+				rel = &r
+				break
+			}
+		}
+	}
+
+	if rel == nil {
+		return fmt.Errorf("could not find Debian image %s", req.Version)
+	}
+
+	tmpZip := paths.New(req.GetTempPath(), "arduino-unoq-debian-image-"+rel.Version+".tar.zst")
+	defer func() { _ = tmpZip.RemoveAll() }()
+
+	if err := updater.DownloadFile(ctx, tmpZip, rel, downloadCB, downloader.Config{}); err != nil {
+		return err
+	}
```

Comment:
This is duplicated code with the `Download` service. Please, move the duplicated code in a private function that is called from both `Download` and `Flash` services.

Moreover, in the `Download` service, the check for free space is not performed. This will be solved automatically when the duplicated code is factored out.